Add Loki as a log exporter (first PR: overall structure) (#1900)
open-telemetry/opentelemetry-collector-contrib: exporter/lokiexporter/README.md

# Loki Exporter

> Exports data via HTTP to [Loki](https://grafana.com/docs/loki/latest/). By default, this exporter requires TLS and

Reviewer: Can an exporter require TLS? I guess it's servers that require it, not exporters. Here it will be TLS or not based on whether endpoint is https or http I think. Or if plaintext really isn't supported, then can you add to the docs for `endpoint` that the URL must start with `https://`?

Author: After a bit more thought, I have removed the requirements for https by default, since some Loki instances my be colocated in the same network as the Otel collector and organizations may not require TLS within the same network.

Updating docs now.

offers queued retry capabilities.

Supported pipeline types: logs

## Getting Started

The following settings are required:

- `endpoint` (no default): The target URL to send Loki log streams to (
  e.g.: https://loki.example.com:3100/loki/api/v1/push).

> - `allowed_attributes_for_stream_labels` (no default): List of allowed attributes to be added as labels to Loki log 

Reviewer: This is so long even the example doesn't use it :) How about just `attributes_for_labels` or similar? Also should these be regexes for more flexibility?

Author: Absolutely agree on the length and the use of regular expressions. Updating now.

  streams. This is a safety net to help prevent accidentally adding dynamic labels that may significantly increase 
  cardinality, thus having a performance impact on your Loki instance. See the 
  [Loki label best practices](https://grafana.com/docs/loki/latest/best-practices/current-best-practices/) page for 
  additional details on the types of labels you may want to associate with log streams.

The following settings can be optionally configured:

- `insecure` (default = false): When set to true disables verifying the server's certificate chain and host name. The
  connection is still encrypted but server identity is not verified.

> - `ca_file` (no default) Path to the CA cert. For a client this verifies the server certificate. Should only be used if

Reviewer: ```suggestion
- `ca_file` (no default) Path to the CA cert to verify the server being connected to.
```

Author: Addressed.

- `cert_file` (no default) Path to the TLS cert to use for client connections when TLS client auth is required. 

>   Should only be used if `insecure`

Reviewer: Since these are for client auth, I guess they're independent of the `insecure` flag? `insecure` is to allow TLS even if `ca_file` isn't set I think

Author: That is correct. The `insecure` flag configures the HTTP client to not validate the server certificate, so it wouldn't make sense to provide a certs if validation is disabled. 

Updating docs so that it says "Should only be used if `insecure` is set to false.

> - `cert_file` (no default) Path to the TLS cert to use for TLS required connections. Should only be used if `insecure`

Reviewer: ```suggestion
- `cert_file` (no default) Path to the TLS cert to use for client connections when TLS client auth is required. Should only be used if `insecure`
```

Author: Addressed

  is set to false.

> - `key_file` (no default) Path to the TLS key to use for TLS required connections. Should only be used if `insecure` is

Reviewer: Ditto

Author: Addressed

  set to false.

- `timeout` (default = 30s): HTTP request time limit. For details see https://golang.org/pkg/net/http/#Client
- `read_buffer_size` (default = 0): ReadBufferSize for HTTP client.
- `write_buffer_size` (default = 512 * 1024): WriteBufferSize for HTTP client.

- `headers` (no default): Name/value pairs added to the HTTP request headers. Loki by default uses the "X-Scope-OrgID"
  header to identify the tenant the log is associated to.

Example:

```yaml
loki:
  endpoint: https://loki.example.com:3100/loki/api/v1/push
  allowed_labels:

>     - app

Reviewer: These are supposed to be attribute names I think so using realizstic semantic convention names would be clearer.

Author: Addressed
